### PR TITLE
Implement getSuggestedRoutes endpoint in the Send Flow for wallet

### LINKED ISSUE #18201: Load suggested routes on the Send flow
# Problem

As a user I need to fetch suggested routes so we fetch the most optimal routes for a transaction

## Implementation

Implement getSuggestedRoutes endpoint from status-go

## Acceptance criteria

No specific UI is required for this issue, that will be implemented separately in #16929

### PR DESCRIPTION
**Pull request by briansztamfater**
fixes #18201

### Summary

This PR includes token amount selection screen into the Send flow and implements getSuggestedRoutes endpoint, which based on the response can enable the Confirm button to proceed to Transaction Confirmation Screen (to be implemented in a follow up PR)

https://github.com/status-im/status-mobile/assets/18485527/f5e3b0ac-dd76-488a-9d33-7104c66ee8d3

#### Platforms

- Android
- iOS

##### Functional

- wallet / transactions

### Steps to test

- Open Status
- Open an account
- Select Send option
- Select a token
- Input an address
- Input an amount, and wait until a route is loaded. Should be enough balance for the sending amount + fees, otherwise a route won't be found
- Verify correct behavior of the screen (see demo video above)

status: ready

### CODE REVIEW

File: src/status_im2/contexts/wallet/send/input_amount/view.cljs [hunk 1]
```diff
@@ -122,7 +140,16 @@
...
+         [rn/scroll-view
+          {:content-container-style {:flex-grow       1
+                                     :align-items     :center
+                                     :justify-content :center}}
+          (cond loading-suggested-routes?
+                [quo/text "Loading routes"]
+                (and (not loading-suggested-routes?) route)
+                [quo/text "Route found"]
+                (and (not loading-suggested-routes?) (nil? route))
+                [quo/text "Route not found"])]
```
Comment (briansztamfater) [inline src/status_im2/contexts/wallet/send/input_amount/view.cljs:149]: This is temporary, should be addressed in #16929

File: src/status_im2/constants.cljs [hunk 2]
```diff
@@ -390,3 +390,15 @@
 (def ^:const account-default-customization-color :blue)
 
 (def ^:const wallet-account-name-max-length 20)
+
+(def ^:const gas-rate-low 0)
```
Comment (J-Son89) [inline src/status_im2/constants.cljs:398]: 🙌

File: src/quo/components/wallet/token_input/component_spec.cljs [hunk 3]
```diff
@@ -6,14 +6,14 @@
...
-               {:token      :snt
+               {:token      {:symbol "snt"}
                 :currency   :eur
                 :conversion 1}])
     (h/is-truthy (h/get-by-text "SNT")))
 
   (h/test "Amount renders"
     (h/render [token-input/view
-               {:token      :snt
+               {:token      {:symbol "snt"}
```
Comment (J-Son89) [inline src/quo/components/wallet/token_input/component_spec.cljs]: why is the apit not just `:token      "snt"`, seems like we could avoid making this component aware of this data structure ?

File: src/status_im2/contexts/wallet/send/input_amount/view.cljs [hunk 4]
```diff
@@ -47,71 +48,88 @@
...
-                                        current-limit-amount (:amount @current-limit)]
-                                    (if (> num-value current-limit-amount)
-                                      (reset! input-value (str current-limit-amount))
-                                      (reset! input-value v))
-                                    (reagent/flush))))]
+  [{:keys [rate]}]
+  (let [bottom                    (safe-area/get-bottom)
+        {:keys [currency]}        (rf/sub [:profile/profile])
+        networks                  (rf/sub [:wallet/network-details])
+        wallet-send               (get-in (rf/sub [:wallet]) [:ui :send])
```
Comment (J-Son89) [inline src/status_im2/contexts/wallet/send/input_amount/view.cljs]: should we not just make a sub for wallet-send?

File: src/status_im2/contexts/wallet/send/input_amount/view.cljs [hunk 5]
```diff
@@ -47,71 +48,88 @@
...
+        token-symbol              (:symbol token)
+        limit-crypto              (:total-balance token)
+        limit-fiat                (:total-balance-fiat token)
+        conversion-rate           (or rate 10)
+        input-value               (reagent/atom "")
+        current-limit             (reagent/atom {:amount   limit-crypto
+                                                 :currency token-symbol})
+        loading-suggested-routes? (:loading-suggested-routes? wallet-send)
+        handle-swap               (fn [crypto?]
+                                    (let [num-value (parse-double @input-value)]
```
Comment (J-Son89) [inline src/status_im2/contexts/wallet/send/input_amount/view.cljs:66]: parse-double is okay to use here? 
cc @ulisesmac

File: src/status_im2/contexts/wallet/send/input_amount/view.cljs [hunk 6]
```diff
@@ -47,71 +48,88 @@
...
-            input-num-value   (parse-double @input-value)
-            confirm-disabled? (or
-                               (empty? @input-value)
-                               (<= input-num-value 0)
-                               (> input-num-value (:amount @current-limit)))]
+          :or   {on-confirm #(rf/dispatch [:wallet/send-select-amount @input-value
+                                           :wallet-send-input-amount])}}]
+      (let [limit-label               (make-limit-label @current-limit)
+            input-num-value           (parse-double @input-value)
+            route                     (get-in (rf/sub [:wallet]) [:ui :send :route])
```
Comment (J-Son89) [inline src/status_im2/contexts/wallet/send/input_amount/view.cljs]: we should make a wallet send sub :)

File: src/utils/money.cljs [hunk 7]
```diff
@@ -205,10 +205,14 @@
       (with-precision 2)
       str))
 
-(defn add
+(defn- add*
   [bn1 n2]
   (.add ^js bn1 n2))
```
Comment (J-Son89) [inline src/utils/money.cljs]: not for now, but these methods should probably all be covered in tests. We can add this as part of the post 1.27 workset 👍

File: src/status_im2/subs/wallet/wallet.cljs [hunk 8]
```diff
@@ -48,7 +48,7 @@
  :<- [:wallet/balances]
  :<- [:wallet/tokens-loading?]
  (fn [[accounts balances tokens-loading?]]
-   (mapv (fn [{:keys [color address type] :as account}]
+   (mapv (fn [{:keys [color address] :as account}]
```
Comment (J-Son89) [inline src/status_im2/subs/wallet/wallet.cljs]: why did you remove type? it's needed on line 54?

File: src/status_im2/contexts/wallet/send/events.cljs [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+                              from-address
+                              to-address
+                              (money/to-hex amount-in)
+                              token-id
+                              []
+                              []
+                              network-preferences
+                              gas-rates
+                              {}]
+         timestamp           (datetime/timestamp)]
```
Comment (smohamedjavid) [inline src/status_im2/contexts/wallet/send/events.cljs]: The current timestamp is available on `now` cofx.

File: src/status_im2/contexts/wallet/send/events.cljs [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+            (update-in [:wallet :ui :send] dissoc :suggested-routes)
+            (update-in [:wallet :ui :send] dissoc :route)
+            (update-in [:wallet :ui :send] dissoc :loading-suggested-routes?))}))
+
+(rf/reg-event-fx :wallet/select-send-address
+ (fn [{:keys [db]} [address]]
+   {:db (assoc-in db [:wallet :ui :send :to-address] address)}))
+
+(rf/reg-event-fx :wallet/send-select-token
+ (fn [{:keys [db]} [token stack-id]]
```
Comment (J-Son89) [inline src/status_im2/contexts/wallet/send/events.cljs]: imo the second parameter here would be better as a map as it's not at all obvious why/what the stack-id is when you're passing it in 👍

File: src/status_im2/contexts/wallet/send/events.cljs [hunk 11]
```diff
@@ -0,0 +1,78 @@
...
+    :fx [[:navigate-to-within-stack [:wallet-send-input-amount stack-id]]]}))
+
+(rf/reg-event-fx :wallet/get-suggested-routes
+ (fn [{:keys [db]} [amount]]
+   (let [wallet-address      (get-in db [:wallet :current-viewing-account-address])
+         token               (get-in db [:wallet :ui :send :token])
+         to-address          (get-in db [:wallet :ui :send :to-address])
+         token-decimal       (:decimals token)
+         token-id            (:symbol token)
+         network-preferences [constants/mainnet-chain-id] ; TODO: don't hardcode network preferences
```
Comment (J-Son89) [inline src/status_im2/contexts/wallet/send/events.cljs]: can we remove this? the network preferences are currently available and by default will be all valid networks anyway 👍

File: src/status_im2/contexts/wallet/send/events.cljs [hunk 12]
```diff
@@ -0,0 +1,78 @@
...
+            (update-in [:wallet :ui :send] dissoc :suggested-routes)
+            (update-in [:wallet :ui :send] dissoc :route)
+            (assoc-in [:wallet :ui :send :loading-suggested-routes?] false))}))
+
+(rf/reg-event-fx :wallet/clean-suggested-routes
+ (fn [{:keys [db]}]
+   {:db (-> db
+            (update-in [:wallet :ui :send] dissoc :suggested-routes)
+            (update-in [:wallet :ui :send] dissoc :route)
+            (update-in [:wallet :ui :send] dissoc :loading-suggested-routes?))}))
```
Comment (smohamedjavid) [inline src/status_im2/contexts/wallet/send/events.cljs:37]: I believe the `dissoc` method can take multiple keys to remove.

File: src/status_im2/contexts/wallet/send/events.cljs [hunk 13]
```diff
@@ -0,0 +1,78 @@
...
+
+(rf/reg-event-fx :wallet/get-suggested-routes
+ (fn [{:keys [db]} [amount]]
+   (let [wallet-address      (get-in db [:wallet :current-viewing-account-address])
+         token               (get-in db [:wallet :ui :send :token])
+         to-address          (get-in db [:wallet :ui :send :to-address])
+         token-decimal       (:decimals token)
+         token-id            (:symbol token)
+         network-preferences [constants/mainnet-chain-id] ; TODO: don't hardcode network preferences
+         gas-rates           constants/gas-rate-low
```
Comment (J-Son89) [inline src/status_im2/contexts/wallet/send/events.cljs]: what's the deal with selecting gas-rate in the UI? when do we select a higher gas rate?

File: src/status_im2/contexts/wallet/send/events.cljs [hunk 14]
```diff
@@ -0,0 +1,78 @@
...
+         amount-in           (money/mul (money/bignumber amount)
```
Comment (J-Son89) [inline src/status_im2/contexts/wallet/send/events.cljs]: might be better to make amount-in a helper fn?
```
(defn amount-in-hex [amount token-decimal]
 (money/to-hex-in (money/mul (money/bignumber amount) (money/from-decimal token-decimal))) 
```

File: src/status_im2/contexts/wallet/common/utils.cljs [hunk 15]
```diff
@@ -43,12 +43,15 @@
...
+  (* (total-token-value-in-all-chains token)
+     (-> token :market-values-per-currency :usd :price)))
+
 (defn calculate-balance
   [tokens-in-account]
   (->> tokens-in-account
-       (map (fn [token]
-              (* (total-token-value-in-all-chains token)
-                 (-> token :market-values-per-currency :usd :price))))
+       (map #(calculate-balance-for-token %))
```
Comment (smohamedjavid) [inline src/status_im2/contexts/wallet/common/utils.cljs:86]:
```suggestion
       (map calculate-balance-for-token)
```

File: src/status_im2/contexts/wallet/send/events.cljs [hunk 16]
```diff
@@ -0,0 +1,78 @@
...
+                                        (money/from-decimal token-decimal))
+         from-address        wallet-address
+         request-params      [constants/send-type-transfer
+                              from-address
+                              to-address
+                              (money/to-hex amount-in)
+                              token-id
+                              []
```
Comment (J-Son89) [inline src/status_im2/contexts/wallet/send/events.cljs]: although empty, might be better to give these names that match the api for easier understanding etc 👍

**Review by J-Son89 — APPROVED** [review]

File: src/status_im2/contexts/wallet/send/select_asset/view.cljs [hunk 17]
```diff
@@ -18,10 +18,13 @@
 (defn- asset-component
   []
   (fn [token _ _ _]
-    (let [on-press                #(js/alert "Not implemented yet")
+    (let [on-press
+          #(if (= (:symbol token) "ETH")
```
Comment (J-Son89) [inline src/status_im2/contexts/wallet/send/select_asset/view.cljs]: is there any reason we need to do this for the moment? does the app crash otherwise? if it's to prevent the app crashing perhaps we should just handle that instead of this ? np either way for me as I realise the updates are coming v soon!

Comment (briansztamfater) [reply src/status_im2/contexts/wallet/send/select_asset/view.cljs]: We can remove this I think, there was an issue with the sendTransaction endpoint, but I can re-check it on the other PR

File: src/status_im2/contexts/wallet/send/events.cljs (hunk 11 above)
Comment (briansztamfater) [reply src/status_im2/contexts/wallet/send/events.cljs]: Correct me if I'm wrong, but this value should be based on network preferences from the receiver? Not sure

Comment (J-Son89) [reply src/status_im2/contexts/wallet/send/events.cljs]: Ah okay, you're probably right. I need to check about that too

File: src/quo/components/wallet/token_input/component_spec.cljs (hunk 3 above)
Comment (ulisesmac) [reply src/quo/components/wallet/token_input/component_spec.cljs]: With changes made in the token component PR (#18018) it no longer needs  that `:symbol` map

File: src/quo/components/wallet/token_input/view.cljs [hunk 18]
```diff
@@ -43,7 +43,7 @@
                      :align-items    :flex-end}}
          [rn/image
           {:style  style/token
-           :source (resources/get-token token)}]
+           :source (resources/get-token (keyword (string/lower-case token)))}]
```
Comment (ulisesmac) [inline src/quo/components/wallet/token_input/view.cljs]: This call will also be simplified, Actually you'll have a conflict here since now we use the token component instead of an image and a `get-token` call

File: src/status_im2/contexts/wallet/common/utils.cljs [hunk 19]
```diff
@@ -44,12 +44,15 @@
     (when balance-in-chain
       (calculate-raw-balance (:raw-balance balance-in-chain) decimals))))
 
+(defn calculate-balance-for-token
+  [token]
+  (* (total-token-value-in-all-chains token)
+     (-> token :market-values-per-currency :usd :price)))
```
Comment (ulisesmac) [inline src/status_im2/contexts/wallet/common/utils.cljs]: Finally merged the bignumber calculations PR, in that PR we calculate the USD price with a utils.money function. I'd suggest to also use it :+1:

File: src/status_im2/contexts/wallet/common/utils.cljs (hunk 15 above)
Comment (ulisesmac) [reply src/status_im2/contexts/wallet/common/utils.cljs:86]: This function has also changes, so some conflicts might arise

File: src/status_im2/contexts/wallet/send/events.cljs (hunk 12 above)
Comment (ulisesmac) [reply src/status_im2/contexts/wallet/send/events.cljs:37]: Yep! @smohamedjavid we can dissoc many keys in a single call

File: src/status_im2/contexts/wallet/send/events.cljs (hunk 14 above)
Comment (ulisesmac) [reply src/status_im2/contexts/wallet/send/events.cljs]: I think the same, as a named function it'll be clearer what these calculations mean :+1:

File: src/status_im2/contexts/wallet/send/input_amount/component_spec.cljs [hunk 20]
```diff
@@ -17,7 +17,13 @@
...
                              :network-name     :ethereum
                              :chain-id         1
-                             :related-chain-id 5}]})
+                             :related-chain-id 5}]
+   :wallet                 {:ui {:send {:token {:symbol                    "ETH"
+                                                :decimals                  18
+                                                :total-balance             1
+                                                :total-balance-fiat        10
+                                                :loading-suggested-routes? false
+                                                :route                     {}}}}}})
```
Comment (ulisesmac) [inline src/status_im2/contexts/wallet/send/input_amount/component_spec.cljs]: :+1:

File: src/status_im2/contexts/wallet/send/input_amount/view.cljs [hunk 21]
```diff
@@ -47,71 +48,88 @@
...
-                                  (let [num-value            (or (parse-double v) 0)
-                                        current-limit-amount (:amount @current-limit)]
-                                    (if (> num-value current-limit-amount)
-                                      (reset! input-value (str current-limit-amount))
-                                      (reset! input-value v))
-                                    (reagent/flush))))]
+  [{:keys [rate]}]
+  (let [bottom                    (safe-area/get-bottom)
+        {:keys [currency]}        (rf/sub [:profile/profile])
+        networks                  (rf/sub [:wallet/network-details])
```
Comment (ulisesmac) [inline src/status_im2/contexts/wallet/send/input_amount/view.cljs:55]: Just keep in mind if a value in these subs change, it won't trigger a rerender

File: src/status_im2/contexts/wallet/send/input_amount/view.cljs (hunk 5 above)
Comment (ulisesmac) [reply src/status_im2/contexts/wallet/send/input_amount/view.cljs:66]: I'd say since this is a user-focused input, we are not going to handle many digits, but we should stablish a limit on how many digits are we going to support. We are dealing with money, so it's better to make sure we properly handle amounts.

IMO the important part comes just before sending any amount to status-go, we must make sure the value is correct (so for that case I think bignumbers are better)

I did a small example to check if doubles will give us what we want (@briansztamfater you could modify it to test expected values)

```clojure
(let [generator (cycle (range 1 10))
      num-1     (take 17 generator)
      num-2     (take 18 generator)
      num-3     (take 20 generator)
      num-4     (take 22 generator)
      nums      [num-1 num-2 num-3 num-4]]
  (as-> nums $
    (map #(apply str %) $)
    (map vector $ (map parse-double $))
    (zipmap '[num-1 num-2 num-3 num-4] $)))
;; =>
 {num-1 ["12345678912345678"      12345678912345678],
  num-2 ["123456789123456789"     123456789123456780],
  num-3 ["12345678912345678912"   12345678912345680000N],
  num-4 ["1234567891234567891234" 1.234567891234568E21]}
```

without considering cases using a `.`, the double loses precision at 18 digits

So please check if parse-double can do the job here

File: src/status_im2/subs/wallet/wallet.cljs (hunk 8 above)
Comment (ulisesmac) [reply src/status_im2/subs/wallet/wallet.cljs]: Maybe we should destructure this map in another way, so that the app doesn't compile if the var doesn't exist.

```clojure
(mapv (fn [{:keys        [color address]
            account-type :type
            :as          account}]
        ,,,
        ;;line 54:
        :type (if (= account-type :watch) :watch-only :empty)
        ,,,)
      [,,,])
```

currently the app is compiling but in the line 54: `(= type :watch)` is always going to be `false`.
Just to give a little more of context on it in case it's needed:
```clojure
(= type :watch)
;; => false

type
;; => #object[cljs$core$type]
```

File: src/status_im2/subs/wallet/wallet.cljs [hunk 22]
```diff
@@ -75,7 +75,7 @@
                        (assoc token
                               :networks           (utils/network-list token networks)
                               :total-balance      (utils/total-token-value-in-all-chains token)
-                              :total-balance-fiat (utils/calculate-balance token)))
+                              :total-balance-fiat (utils/calculate-balance-for-token token)))
```
Comment (ulisesmac) [inline src/status_im2/subs/wallet/wallet.cljs]: Great it's solved, but since the PR involving bignumber balances has  been merged, I'd suggest using an implementation based on bignumbers :+1:

File: src/utils/money.cljs (hunk 7 above)
Comment (ulisesmac) [reply src/utils/money.cljs]: I added it (and explained why) in a previous PR.

https://github.com/status-im/status-mobile/pull/17920#discussion_r1394942416

I can add some tests in a following PR @J-Son89

File: src/status_im2/contexts/wallet/send/events.cljs (hunk 13 above)
Comment (briansztamfater) [reply src/status_im2/contexts/wallet/send/events.cljs]: I didn't find a design in Figma where the user could customize this, but in general I think it is fine as a first approach to have low gas rate as default, then user could customize in some advanced settings. We should definitely ask design team the correct approach so we can track some follow up issue for this 👍

File: src/status_im2/subs/wallet/wallet.cljs (hunk 22 above)
Comment (briansztamfater) [reply src/status_im2/subs/wallet/wallet.cljs]: thanks, updated!

File: src/status_im2/contexts/wallet/send/input_amount/view.cljs (hunk 5 above)
Comment (briansztamfater) [reply src/status_im2/contexts/wallet/send/input_amount/view.cljs:66]: I had no major issues while testing, but I agree we should use a safer approach using bignumbers. I made no changes to this code, so I would prefer to handle this in a separate issue! I will create it

File: src/status_im2/contexts/wallet/send/events.cljs (hunk 13 above)
Comment (J-Son89) [reply src/status_im2/contexts/wallet/send/events.cljs]: Desktop use medium as the default, let's use the same 👍

File: src/utils/money.cljs (hunk 7 above)
Comment (J-Son89) [reply src/utils/money.cljs]: @ulisesmac - I was making no comment about the added function, I understand the use case. My point is more that we are using a bunch of (afaik) untested utilities that handle numbers. Tests would be great but we can wait until post 1.27 as let's focus on the delivering those features to the best quality as possible :)

File: src/status_im2/contexts/wallet/send/events.cljs [hunk 23]
```diff
@@ -0,0 +1,85 @@
...
+(rf/reg-event-fx :wallet/clean-suggested-routes
+ (fn [{:keys [db]}]
+   {:db (-> db
+            (update-in [:wallet :ui :send] dissoc :suggested-routes)
+            (update-in [:wallet :ui :send] dissoc :route)
+            (update-in [:wallet :ui :send] dissoc :loading-suggested-routes?))}))
+
+(rf/reg-event-fx :wallet/select-send-address
+ (fn [{:keys [db]} [{:keys [address stack-id]}]]
+   (println "dsadadasd" address stack-id)
```
Comment (vkjr) [inline src/status_im2/contexts/wallet/send/events.cljs]: leftover)

File: src/status_im2/contexts/wallet/send/events.cljs [hunk 24]
```diff
@@ -0,0 +1,85 @@
...
+    :fx [[:navigate-to-within-stack [:wallet-select-asset stack-id]]]}))
+
+(rf/reg-event-fx :wallet/send-select-token
+ (fn [{:keys [db]} [{:keys [token stack-id]}]]
+   {:db (assoc-in db [:wallet :ui :send :token] token)
+    :fx [[:navigate-to-within-stack [:wallet-send-input-amount stack-id]]]}))
+
+(rf/reg-event-fx :wallet/send-select-amount
+ (fn [{:keys [db]} [{:keys [amount]}]]
+   (js/alert "Not implemented yet")
```
Comment (vkjr) [inline src/status_im2/contexts/wallet/send/events.cljs:51]: Wondering if there is an :fx for `js/alert`, if so - it would be more canonical to use it

**Review by vkjr — APPROVED** [review]

File: src/status_im2/contexts/wallet/send/events.cljs (hunk 24 above)
Comment (briansztamfater) [reply src/status_im2/contexts/wallet/send/events.cljs:51]: Not that I am aware of, but would be great to add it!

File: src/utils/money.cljs (hunk 7 above)
Comment (ulisesmac) [reply src/utils/money.cljs]: Oh yes, that's true!, we are relying on the functions in this namespace. They need to be tested.

File: src/utils/money.cljs [hunk 25]
```diff
@@ -246,3 +246,7 @@
 (schema/=> format-amount
   [:=> [:cat [:maybe :int]]
    [:maybe :string]])
+
+(defn amount-in-hex
```
Comment (OmarBasem) [inline src/utils/money.cljs:250]: A test can be added for this function

**Comment by VolodLytvynenko** [comment]
Hey. @briansztamfater the routes are not found in my cases. But I have some real tokens on mainnet, optimizm, arbitrium networks. What I am doing wrong?

### Actual result:

https://github.com/status-im/status-mobile/assets/52490791/fc71c184-ab99-4dff-9e16-6fd5b6699625

### Expected result:
Route is found

### Logs:
[Status-debug-logs.zip](https://github.com/status-im/status-mobile/files/13712785/Status-debug-logs.zip)

**Comment by VolodLytvynenko** [comment]
**Question 1**: Should this feature work in case of founding the route for testnet tokens? For now it does not work

### Steps:
1. Recover user with Goerli tokens
2. Enable testnet mode
3. Go to 'send to' page
4. Enter an available amount of tokens for transfer

### Actual result:
The route is not found

https://github.com/status-im/status-mobile/assets/52490791/51cab9b1-8c6f-4d0a-9ba1-1adb8f7eb25c


### Expected result:
Should the rout be found?

**Comment by J-Son89** [comment]
Hi @VolodLytvynenko, that should at least work. This is the happy path for finding the best route for a transaction. I tested yesterday evening and it was working okay for me. Is it working on mainnet?

**Comment by VolodLytvynenko** [comment]
> Hi @VolodLytvynenko, that should at least work. This is the happy path for finding the best route for a transaction. I tested yesterday evening and it was working okay for me. Is it working on mainnet?

@J-Son89 Yes, I tried it on the Mainnet network without any success to get the route, even when entering the minimum amount.
Here's the balance from my account that I used:

image

**Comment by VolodLytvynenko** [comment]
@J-Son89 Have I understood the requirements correctly regarding route finding within the scope of this PR?

The route should always be found if the user has enough assets on all networks. For instance:
**Given** a user has the following assets:
1 ETH on the Optimism network
2 ETH on the Mainnet network
**When** the user goes to the 'send to' page and enters 3 ETH
**Then** the route should always be found.

However, in the case of entering an amount higher than available, for example, '3.0001 ETH,' the route should not be found.

 Is there a minimum amount for a particular asset that determines whether a route should or should not be found?


Or in the scope of this PR would be enough to see the route is found at least once to achieve the happy path?

**Comment by J-Son89** [comment]
> @J-Son89 Have I understood the requirements correctly regarding route finding within the scope of this PR?
> 
> The route should always be found if the user has enough assets on all networks. For instance: **Given** a user has the following assets: 1 ETH on the Optimism network 2 ETH on the Mainnet network **When** the user goes to the 'send to' page and enters 3 ETH **Then** the route should always be found.
> 
> However, in the case of entering an amount higher than available, for example, '3.0001 ETH,' the route should not be found.
> 
> Is there a minimum amount for a particular asset that determines whether a route should or should not be found?
> 
> Or in the scope of this PR would be enough to see the route is found at least once to achieve the happy path?

Hey @VolodLytvynenko -
this pr adds the most basic support for route finding and at that it is the happy path for this.

i.e It is mostly just for Ethereum tokens (other tokens should work too), It is only a single token transaction, i.e you are sending eth so you will then pay gas in eth and not in other tokens.
In this case it is just a happy path to find that a route has been found.

The amount entered is in crypto, we will handle the fiat conversion in another issue too.

**Comment by VolodLytvynenko** [comment]
@briansztamfater @J-Son89 No issues from my side. The happy flow works. Follow ups I will add separately Ready to be merged
